### ISSUE TITLE
private_cluster_config block missing from state in resource google_container_cluster when public nodes but master cidr defined.

### ISSUE THREAD

**Issue by slancio** [terraform-providers/terraform-provider-google#4773]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment
* If an issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to "hashibot", a community member has claimed the issue already.

### Terraform Version
Terraform v0.12.8

Tested and affected with provider.google versions 2.12 through 2.18.1

### Affected Resource(s)

* google_container_cluster

### Terraform Configuration Files

```tf
resource "google_container_cluster" "this" {
  [...]
  private_cluster_config {
    enable_private_endpoint = false
    enable_private_nodes = false
    master_ipv4_cidr_block = "10.255.240.16/28" # for example
  }
}
```

### Expected Behavior

State for this block should be stored in the terraform state.  Changes to the resource configuration should not cause a provider error when this block is defined.

### Actual Behavior

The private_cluster_config block is completely absent from the terraform state when running terraform state show.  Any changes to the configuration after the first apply cause the terraform provider to throw an error:

```
Error: Provider produced invalid plan

Provider "google" has indicated "requires replacement" on
module.management.google_container_cluster.this for a non-existent attribute
path cty.Path{cty.GetAttrStep{Name:"private_cluster_config"},
cty.IndexStep{Key:cty.NumberIntVal(0)},
cty.GetAttrStep{Name:"master_ipv4_cidr_block"}}.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

This seems to be because we want the nodes/endpoint to be public but specify the master cidr block to be a specific range.

### Steps to Reproduce

1.  Define a google_container_cluster resource for a public cluster with a predefined master_ipv4_cidr_block.
2. `terraform apply`
3. `terraform show` or `terraform state show google_container_cluster.this` (to see that the private_cluster_config block is absent)
4. Change a setting, for example the maintenance window time (to see that terraform errors on changes).

### Important Factoids

We are defining regional clusters attached to a subnetwork with predefined pod & service ranges and a master_ipv4_cidr_block in a previously-unallocated network range.

### References

* #0000

**Comment by slancio** [comment]
Realized that my initial problem definition was inaccurate and revised to clarify.

Shouldn't it be reasonable to define the master_ipv4_cidr_block for public clusters as well as private ones?  Would it make sense to pull that field out of the private_cluster_config block?  It would be nice if the block weren't omitted entirely from the state as that's the only place that would show the private endpoint IP for a public cluster.

**Comment by rileykarson** [comment]
While [private clusters have both a public and private endpoint](https://cloud.google.com/kubernetes-engine/docs/how-to/private-clusters#endpoints), public clusters have a sole public endpoint. Clusters gain a private endpoint when their nodes are private- when `enable_private_nodes` is set to `true`. Since `master_ipv4_cidr_block` controls the ip range of the private master endpoint, it can't be set when `enable_private_nodes` is set to `false`.

Per your second comment- we didn't put the field in the block, it's like that in the underlying API because the feature only works for the private endpoint. Normally, your request to the API would fail. However, based on old issues with the field during Terraform `0.11`, we added a suppression function that accidentally caused the entire block to not get sent when `enable_private_nodes` is `false` at creation time.

While I'd really like Terraform to fail on invalid requests here, it's difficult/impossible to do without a breaking change for some other users of the field. The overzealous suppression function is hard to replicate in a compatible way. I'll amend the documentation to make this more clear & file an issue to fix this for good in `4.0.0`.

**Comment by slancio** [comment]
Ahh, that makes things much more clear.  Thank you for the clarification -- I'll update my terraform code accordingly.
